app/main.py

```diff
@@ -1,2 +1,103 @@
+from typing import Any, Hashable, Iterator
+from collections import namedtuple
+
+
+Node = namedtuple("Node", ["key", "value", "hash_value"])
+
+
 class Dictionary:
-    pass
+    def __init__(self, capacity: int = 8, load_factor: float = 0.67) -> None:
+        self._capacity = capacity
+        self._load_factor = load_factor
+        self.length = 0
+        self.hash_table = [None] * self._capacity
+
+    def __setitem__(self, key: Hashable, value: Any) -> None:
+        hash_value = hash(key)
+        index = hash_value % self._capacity
+        for range_index in range(self._capacity):
+            next_index = (index + range_index) % self._capacity
+            node = self.hash_table[next_index]
+            if node is None:
+                self.hash_table[next_index] = Node(key, value, hash_value)
+                self.length += 1
+                if self.length > self._capacity * self._load_factor:
+                    self._resize()
+                return
+            elif node.key == key:
+                new_node = Node(key, value, hash_value)
+                self.hash_table[next_index] = new_node
+                return
+
+    def __getitem__(self, key: Hashable) -> Any:
+        hash_value = hash(key)
+        index = hash_value % self._capacity
+        for range_index in range(self._capacity):
+            next_index = (index + range_index) % self._capacity
+            node = self.hash_table[next_index]
+            if node is None:
+                break
+            if node.key == key:
+                return node.value
+        raise KeyError(f"Key '{key}' not found")
+
+    def __iter__(self) -> Iterator:
+        for node in self.hash_table:
+            if node is not None:
+                yield node.key
+
+    def __len__(self) -> int:
+        return self.length
+
+    def __delitem__(self, key: Hashable) -> None:
+        hash_value = hash(key)
+        index = hash_value % self._capacity
+        for range_index in range(self._capacity):
+            next_index = (index + range_index) % self._capacity
+            node = self.hash_table[next_index]
+            if node is None:
+                raise KeyError(f"Key '{key}' not found")
+            if node.key == key:
+                self.hash_table[next_index] = None
+                self.length -= 1
+                self._rehash()
```

Comment:
no need to rehash after deleting item

Reply:
The rehashing is still necessary because the deletion may cause the hash table to become unbalanced, and this can lead to hash collisions.

For example, consider a hash table where all keys have the same hash value, and several of them are deleted. In this case, many slots in the hash table will be set to None, and the remaining key-value pairs will cluster together in a small portion of the hash table. This can cause long search times and hash collisions, which can severely degrade performance.

By rehashing the hash table, we can distribute the remaining key-value pairs more evenly across the hash table, reducing the chance of collisions and improving performance.

Reply:
Agree, but I think you should change `_rehash()` method implementation, you create a new instance of the dictionary but in the result, you just store attributes from it. You already have the method with similar logic `_resize()` but with capacity changing. Try to unite these both methods into one universal for both cases